slapo/schedule.py

```diff
@@ -864,27 +880,25 @@ def consolidate_model(sch: Schedule, topology=None):
             )
 
         # use original shape to initialize parameters
-        if global_rank == curr_stage_devices[0] and not "DEBUG" in os.environ:
+        if global_rank == curr_stage_devices[0] and num_params > 0:
             # only the first device in the PP group needs to initialize the weights
-            if hasattr(sch.mod, "_init_weights"):
+            if param_init_fn:
+                param_init_fn(sch.mod)
+            elif hasattr(sch.mod, "_init_weights"):
                 # `_init_weights` is a HF specific API, see
                 # https://github.com/huggingface/transformers/blob/v4.25.1/src/transformers/models/bert/modeling_bert.py#L748
                 sch.mod._init_weights(sch.mod)
             elif hasattr(sch.mod, "reset_parameters"):
                 sch.mod.reset_parameters()
             else:
                 raise RuntimeError(
-                    f"Module {sch.name} should have `reset_parameters` or `_init_weights` method in order to support delay initialization"
+                    f"Module {sch.name} should have `reset_parameters` or `_init_weights` method or param_init_fn={param_init_fn} needs to be provided in order to support delay initialization"
                 )
 
         # need to broadcast params from rank 0 to make sure all the TP+DP ranks take the same params
-        if not "DEBUG" in os.environ:
-            curr_stage_group = stage_groups[curr_part_idx]
-            for _, param in sch.mod.named_parameters(recurse=False):
-                dist.broadcast(param, src=curr_stage_devices[0], group=curr_stage_group)
-            # destroy process group to free memory
-            for g in stage_groups:
-                dist.destroy_process_group(g)
```

Comment:
Is the speed-up from the deletion of this line?

Reply:
no, this causes nccl error